### ISSUE TITLE
Run a script for a project

### ISSUE THREAD

**Issue by bpavie** [qupath/qupath#103]
Dear All,

I am trying to perform analysis on multiple image (CZI files) containing multiple scene.
I created a project and loaded the 2 files, so from the interface, it looks like that:
![qpath_multiple_image_multiple_scene_01](https://user-images.githubusercontent.com/1775952/30912322-60af88da-a38c-11e7-9c5f-7cf53aabd4fb.png)

I ran the following script for 2 czi files containing multiple scene in order to detect the tissue:
```setImageType('BRIGHTFIELD_H_DAB');
runPlugin('qupath.imagej.detect.tissue.SimpleTissueDetection2', '{"threshold": 200,  "requestedPixelSizeMicrons": 5.0,  "minAreaMicrons": 10000.0,  "maxHoleAreaMicrons": 1000.0,  "darkBackground": false,  "smoothImage": false,  "medianCleanup": true,  "dilateBoundaries": false,  "smoothCoordinates": true,  "excludeOnBoundary": false,  "singleAnnotation": true}');
setColorDeconvolutionStains('{"Name" : "H-DAB default", "Stain 1" : "Hematoxylin", "Values 1" : "0.65111 0.70119 0.29049 ", "Stain 2" : "DAB", "Values 2" : "0.26917 0.56824 0.77759 ", "Background" : " 255 255 255 "}');
```

It works, it looks the result is saved using the scene name, so in my case, for example, scene 0, only one ROI is saved for the 2 scenes 0:
![qpath_multiple_image_multiple_scene](https://user-images.githubusercontent.com/1775952/30912398-9f537588-a38c-11e7-8b41-44454126f5eb.png)

and:
![image](https://user-images.githubusercontent.com/1775952/30912429-c32073e4-a38c-11e7-9158-d98fdd48d230.png)

So it looks like the ROI is wrongly set . 

Best,
Benjamin

**Comment by petebankhead** [comment]
Hmmm, this isn't a scenario I ever had to deal with myself... it looks like an unfortunate limitation of how the project arranges ```.qpdata``` files simply according to the image name stored for the entry in the image.

There's no 'good' way to fix it currently, unless you're willing to put all your images in separate projects... which would kind of defeat the purpose of using a project.

So you could try this as a workaround:

```groovy
guiscript=true

// Get QuPath & project
def qupath = getQuPath()
def project = qupath.getProject()

// Loop through images, setting the name
// (actually accessing a private field... therefore 'bad')
project.getImageList().each {
    def path = it.getServerPath()
    int ind = path.lastIndexOf(':')
    def scene = path[ind+1..-1]
    def name = new File(path[0..ind-2]).getName()
    it.putMetadataValue('Slide_ID', name)
    it.imageName = name + ' (' + scene + ')'
    print it.imageName
}

// Need to set to null first to force update
qupath.setProject(null)
qupath.setProject(project)

// Be very careful is you use this to write the project!
// The logic is a bit weird and it will probably overwrite 
// the existing project - so duplicate your .qpproj file to be safer
//qupath.lib.projects.ProjectIO.writeProject(project)
```

Basically, this should rename the images in the project to include both the original file name and the scene.  This should then be used by QuPath when arranging the ```.qpdata``` files afterwards.

It won't automatically update the names of any existing data files - this would have to be done manually.

It has the added bonus of setting the 'Slide_ID' keyword; if you right-click on the project, you can then choose to *Sort by &rarr; Slide ID*.

**Comment by bpavie** [comment]
Thanks for the workaround, I'll give it a try!
Benjamin

**Comment by bpavie** [comment]
Thanks Pete, it works, although it generate a multiple times the following error message:

`
ERROR: QuPath exception
    at com.sun.javafx.tk.Toolkit.checkFxUserThread(Toolkit.java:236)
    at com.sun.javafx.tk.quantum.QuantumToolkit.checkFxUserThread(QuantumToolkit.java:423)
    at javafx.scene.Parent$2.onProposedChange(Parent.java:367)
    at com.sun.javafx.collections.VetoableListDecorator.setAll(VetoableListDecorator.java:113)
    at com.sun.javafx.collections.VetoableListDecorator.setAll(VetoableListDecorator.java:108)
    at com.sun.javafx.scene.control.skin.LabeledSkinBase.updateChildren(LabeledSkinBase.java:575)
    at com.sun.javafx.scene.control.skin.TreeCellSkin.updateChildren(TreeCellSkin.java:197)
    at com.sun.javafx.scene.control.skin.LabeledSkinBase.handleControlPropertyChanged(LabeledSkinBase.java:204)
    at com.sun.javafx.scene.control.skin.TreeCellSkin.handleControlPropertyChanged(TreeCellSkin.java:125)
    at com.sun.javafx.scene.control.skin.BehaviorSkinBase.lambda$registerChangeListener$61(BehaviorSkinBase.java:197)
    at com.sun.javafx.scene.control.MultiplePropertyChangeListenerHandler$1.changed(MultiplePropertyChangeListenerHandler.java:55)
    at javafx.beans.value.WeakChangeListener.changed(WeakChangeListener.java:89)
    at com.sun.javafx.binding.ExpressionHelper$SingleChange.fireValueChangedEvent(ExpressionHelper.java:182)
    at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81)
    at javafx.beans.property.StringPropertyBase.fireValueChangedEvent(StringPropertyBase.java:103)
    at javafx.beans.property.StringPropertyBase.markInvalid(StringPropertyBase.java:110)
    at javafx.beans.property.StringPropertyBase.set(StringPropertyBase.java:144)
    at javafx.beans.property.StringPropertyBase.set(StringPropertyBase.java:49)
    at javafx.beans.property.StringProperty.setValue(StringProperty.java:65)
    at javafx.scene.control.Labeled.setText(Labeled.java:145)
    at qupath.lib.gui.panels.ProjectBrowser$ImageEntryCell.updateItem(ProjectBrowser.java:813)
    at javafx.scene.control.TreeCell.updateItem(TreeCell.java:526)
    at javafx.scene.control.TreeCell.lambda$new$256(TreeCell.java:173)
    at javafx.beans.WeakInvalidationListener.invalidated(WeakInvalidationListener.java:83)
    at com.sun.javafx.binding.ExpressionHelper$Generic.fireValueChangedEvent(ExpressionHelper.java:349)
    at com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:81)
    at javafx.beans.property.ObjectPropertyBase.fireValueChangedEvent(ObjectPropertyBase.java:105)
    at javafx.beans.property.ObjectPropertyBase.markInvalid(ObjectPropertyBase.java:112)
    at javafx.beans.property.ObjectPropertyBase.set(ObjectPropertyBase.java:146)
    at javafx.scene.control.TreeView.setRoot(TreeView.java:470)
    at qupath.lib.gui.panels.ProjectBrowser.setProject(ProjectBrowser.java:271)
    at qupath.lib.gui.QuPathGUI.setProject(QuPathGUI.java:4186)
    at qupath.lib.gui.QuPathGUI$setProject$0.call(Unknown Source)
    at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
    at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
    at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
    at Script5.run(Script5.groovy:21)
    at org.codehaus.groovy.jsr223.GroovyScriptEngineImpl.eval(GroovyScriptEngineImpl.java:343)
    at org.codehaus.groovy.jsr223.GroovyScriptEngineImpl.eval(GroovyScriptEngineImpl.java:152)
    at qupath.lib.scripting.DefaultScriptEditor.executeScript(DefaultScriptEditor.java:765)
    at qupath.lib.scripting.DefaultScriptEditor.executeScript(DefaultScriptEditor.java:695)
    at qupath.lib.scripting.DefaultScriptEditor.executeScript(DefaultScriptEditor.java:677)
    at qupath.lib.scripting.DefaultScriptEditor.access$400(DefaultScriptEditor.java:136)
    at qupath.lib.scripting.DefaultScriptEditor$ProjectTask.call(DefaultScriptEditor.java:1267)
    at qupath.lib.scripting.DefaultScriptEditor$ProjectTask.call(DefaultScriptEditor.java:1195)
    at javafx.concurrent.Task$TaskCallable.call(Task.java:1423)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
`

**Comment by petebankhead** [comment]
Hi Benjamin, the
```groovy
guiscript=true
```
line has to be the very first line in the script.  It's a rather hackish, QuPath-specific trick to ensure that the script is run on the one JavaFX thread that is allowed to interact with the GUI.  Otherwise, any attempt to do anything affecting the GUI directly results in the kind of errors you're seeing.

There's an example of its use [here](https://gist.github.com/petebankhead/6f73a01a67935dae2f7fa75fabe0d6ee).

Without the guiscript trick, you'd need to wrap the contents of the script in something like this
```groovy
javafx.application.Platform.runLater {
  // Everything else here...
}
```
